### ISSUE TITLE
Poor performance when accessing nested properties

### ISSUE THREAD

**Issue by tdzienniak** [stampit-org/stampit#269]
I encountered a problem with very poor performance when accessing nested props in a loop.
I created a repo with benchmark: [https://github.com/tdzienniak/stampit-benchmark](https://github.com/tdzienniak/stampit-benchmark)

Screenshot from benchmark result:
![stampit-benchmark](https://cloud.githubusercontent.com/assets/1304915/19809133/3d347a16-9d28-11e6-9f42-d37ea2ed2cea.PNG)

As you can see, the difference is HUGE.

The code is a simple loop updating the position of game object, because I encounterd this issue when developing game engine with stampit. I was trying to figure out why objects created from stamps are so much slower, but I don't have a clue.

**Comment by koresar** [comment]
Good! I need:
1) stampit version
2) stamp source code
3) your benchmark code
:)
Cheers!

**Comment by tdzienniak** [comment]
stampit version is 3.0.4. Stamp and my benchmark code is in the repo: [https://github.com/tdzienniak/stampit-benchmark/blob/master/index.js](https://github.com/tdzienniak/stampit-benchmark/blob/master/index.js)
I'm using this tool for benchmarking: [https://github.com/bestiejs/benchmark.js](https://github.com/bestiejs/benchmark.js)

**Comment by koresar** [comment]
Reproduced. This suck! I found the cause. Sorry about that mate!
Here, I fixed it:
```
> npm start

> stampit-benchmark@1.0.0 start /Users/vasyl/code/stampit-benchmark
> node ./index.js

Running benchmark...
Stampit x 14,445 ops/sec ±1.80% (78 runs sampled)
Plain object x 10,127 ops/sec ±6.99% (77 runs sampled)
Fastest is Stampit
```

Could you please allow me 24-48 hours to fix that? (It's Saturday 1 am here.)

Meanwhile, if you need a workaround see [my PR to your repo](https://github.com/tdzienniak/stampit-benchmark/pull/1).

**Comment by tdzienniak** [comment]
Thank you mate. Take your time. :)

**Comment by koresar** [comment]
It's horrible! The behavior of the fix I found is too different on different node.js versions!

I copied the test case from above to stampit. Wanted to have a benchmark test in stampit. Here are the results:
```sh
┌vasyl@192-168-1-7 ~/c/stampit (master *+$)                                
└> nvm use 6
Now using node v6.9.1 (npm v3.10.8)
┌vasyl@192-168-1-7 ~/c/stampit (master *+$)                                
└> ./node_modules/.bin/babel-node  --presets=es2015 test/benchmark/
TAP version 13
# benchmarking property access
Stampit x 26,613 ops/sec ±1.08% (82 runs sampled)
Plain object x 12,512 ops/sec ±1.30% (82 runs sampled)
ok 1 object instances property access must be as fast as plain object

1..1
# tests 1
# pass  1

# ok
```
```sh
┌vasyl@192-168-1-7 ~/c/stampit (master *+$)                                
└> nvm use 4
Now using node v4.6.0 (npm v2.15.9)
┌vasyl@192-168-1-7 ~/c/stampit (master *+$)                                
└> ./node_modules/.bin/babel-node  --presets=es2015 test/benchmark/
TAP version 13
# benchmarking property access
Stampit x 517 ops/sec ±1.29% (81 runs sampled)
Plain object x 13,979 ops/sec ±1.26% (76 runs sampled)
not ok 1 object instances property access must be as fast as plain object
  ---
    operator: ok
    expected: true
    actual:   false
    at: Suite.<anonymous> (/Users/vasyl/code/stampit/test/benchmark/property-access.js:76:9)
  ...

1..1
# tests 1
# pass  0
# fail  1
```

You read it right! Object instances access is twice faster for stamps, but only in latest node.js (v6 and v7). But in node 4 it's horribly slow regardless!

I'm trying to find the best workaround to fix the thing for all VMs.

**Comment by koresar** [comment]
Okay. Here is the story.

TL;DR V8 runtime optimizations depends on too many things.

=====

Node.js v4 and v5 kick the "class-object" optimization somewhat differently to v6 and v7. They've changed something internally at some point.

So I changed the benchmark a little. Made sure it behaves the same on all node.js versions we support.

I made the benchmark as a tape test under `test/benchmark/`.

I fixed the slow performance issue. Property access should now be almost as fast as plain objects.

The bugfix PR is here: #270
Will publish the new release as soon as PR gets merged.

=====

Thank you very much @tdzienniak for the issue. Consider creating new issues if you find something similar.

**Comment by koresar** [comment]
Travis CI has [these benchmarking results](https://travis-ci.org/stampit-org/stampit/builds/171612960):

v4
`Stampit x 16,229 ops/sec ±0.93% (82 runs sampled). Plain object x 13,175 ops/sec ±1.71% (86 runs sampled)`

v5
`Stampit x 9,167 ops/sec ±3.73% (76 runs sampled). Plain object x 10,101 ops/sec ±4.31% (77 runs sampled)`

v6
`Stampit x 20,973 ops/sec ±0.55% (83 runs sampled). Plain object x 10,199 ops/sec ±9.37% (87 runs sampled)`

v7
`Stampit x 16,355 ops/sec ±10.86% (84 runs sampled). Plain object x 11,150 ops/sec ±2.12% (87 runs sampled)`

The code for the test is [here](https://github.com/stampit-org/stampit/blob/4d104adc4ce4ad1415eb5e2d5d6c54b8f78f67fa/test/benchmark/property-access.js#L9).

**Comment by koresar** [comment]
v3.0.5 was just published. Enjoy!
